File: typescript/src/mode.ts
import * as tty from "node:tty";

/** True if INQUIRER_AI_MODE == "human" (case-insensitive). */
export function isHumanMode(): boolean {
  return (process.env.INQUIRER_AI_MODE ?? "").toLowerCase() === "human";
}

/**
 * A socket transport is requested when INQUIRER_AI_SOCKET is set & non-empty,
 * OR INQUIRER_AI_MODE == "agent" (case-insensitive) (R3).
 */
export function isSocketRequested(): boolean {
  const sock = process.env.INQUIRER_AI_SOCKET;
  if (sock !== undefined && sock !== "") return true;
  return (process.env.INQUIRER_AI_MODE ?? "").toLowerCase() === "agent";
}

/**
 * Agent (non-interactive) mode. is_agent = (not human) AND
 * (socket_requested OR stdin-is-not-a-TTY). A plain piped non-TTY with no
 * MODE/SOCKET stays in agent (stdio) mode for backwards compatibility (R3).
 */
export function isAgentMode(): boolean {
  if (isHumanMode()) return false;
  if (isSocketRequested()) return true;
  return !tty.isatty(0);
}

File: rust/src/mode.rs
use std::env;
use std::io::IsTerminal;

/// Returns true when `INQUIRER_AI_MODE` equals "human" (case-insensitive).
pub fn is_human_mode() -> bool {
    env::var("INQUIRER_AI_MODE")
        .unwrap_or_default()
        .eq_ignore_ascii_case("human")
}

/// Returns true when a socket transport has been explicitly requested:
/// either `INQUIRER_AI_SOCKET` is set and non-empty, or
/// `INQUIRER_AI_MODE` equals "agent" (case-insensitive).
pub fn is_socket_requested() -> bool {
    let socket_set = env::var("INQUIRER_AI_SOCKET")
        .map(|v| !v.is_empty())
        .unwrap_or(false);
    if socket_set {
        return true;
    }
    env::var("INQUIRER_AI_MODE")
        .unwrap_or_default()
        .eq_ignore_ascii_case("agent")
}

/// Agent mode is active when NOT human AND (a socket was requested OR stdin
/// is not a TTY).
///
/// This implements the unified mode-detection contract (R3): a plain piped
/// non-TTY with no MODE/SOCKET stays in agent (stdio) mode for backwards
/// compatibility, while `INQUIRER_AI_SOCKET` (even on a TTY) activates agent
/// mode.
pub fn is_agent_mode() -> bool {
    if is_human_mode() {
        return false;
    }
    is_socket_requested() || !std::io::stdin().is_terminal()
}
